### PR TITLE
HIP: optimize SVDQuant INT4 activation quantization

### PR DESCRIPTION
**Pull request by tangzzycc**
## Summary

This PR improves the HIP preprocessing path used by `quantize_svdquant_w4a4`:

- Process each 64-element INT4 quantization group with 16 cooperating threads.
- Reuse the smoothed activation values for reduction and packing, avoiding a second input read.
- Use `torch.mm` for LoRA-down, matching the existing CUDA backend, and remove the dedicated scalar HIP binding.

The public Python API and output layout remain unchanged.

## Implementation

Sixteen threads process each 64-element quantization group. Each thread loads and retains four smoothed values, participates in a 16-lane shuffle reduction, and writes four packed INT4 values through one 16-bit store. The existing scalar kernel remains available as a fallback for operands with different dtypes.

LoRA-down is dispatched through PyTorch's matrix multiplication path instead of the previous one-output-element-per-thread HIP kernel.

## Performance

Measured on an AMD Radeon 8060S (`gfx1151`) with ROCm 7.2. Each result uses 20 warmup iterations followed by 100 measured iterations and reports the median kernel time.

| dtype | shape | Before | After | Speedup |
|---|---:|---:|---:|---:|
| FP16 | 4096 x 4096 | 3.723 ms | 0.214 ms | 17.4x |
| FP16 | 4096 x 14336 | 15.363 ms | 0.713 ms | 21.5x |
| BF16 | 4096 x 4096 | 3.731 ms | 0.212 ms | 17.6x |
| BF16 | 4096 x 14336 | 15.250 ms | 0.712 ms | 21.4x |
| FP32 | 4096 x 4096 | 2.604 ms | 0.350 ms | 7.4x |
| FP32 | 4096 x 14336 | 11.105 ms | 1.220 ms | 9.1x |

Based on the minimum required input/output traffic, the large FP16/BF16 cases reach approximately 198-209 GB/s of effective bandwidth.

## Architecture verification

The extension was successfully compiled for all currently supported gfx11/gfx12 targets:

```text
gfx1100 gfx1101 gfx1102 gfx1103
gfx1150 gfx1151 gfx1152 gfx1153
gfx1200 gfx1201
```

Code-object inspection confirmed packed quantizer stores and no private-memory spills in the optimized quantization kernel. Runtime performance was measured on gfx1151; the remaining targets were validated through compilation and code-object inspection.

## Validation

```bash
CMAKE_BUILD_PARALLEL_LEVEL=16 .venv/bin/python setup.py \
  build_ext --inplace --no-cuda --hip --hip-archs=gfx1151

.venv/bin/pytest tests/test_svdquant_w4a4.py tests/test_hip_wmma.py -q
```

Result:

```text
370 passed, 14 skipped
```

The tests cover signed and unsigned INT4 packing, quantization-group layout, transposed scales, padded rows, and LoRA-down results against `torch.mm`.

### CODE REVIEW

File: tests/test_hip_wmma.py [hunk 1]
```diff
@@ -1975,6 +1975,72 @@ def test_svdquant_validates_its_operands(hip):
...
+        values = values.remainder(16)
+    else:
+        values = values.remainder(15) - 7
+    x = values.to(dtype)
+    smooth = torch.ones(k, device=DEV, dtype=dtype)
+    lora_down = torch.zeros(k, 1, device=DEV, dtype=dtype)
+
+    q, ascales, _ = hip.quantize_svdquant_w4a4(
+        x, smooth, lora_down, pad_size=16, act_unsigned=act_unsigned
+    )
```
[the remark this answers is not in the document]
Comment (tangzzycc) [reply tests/test_hip_wmma.py:2015]: Addressed. Added K=1088 coverage to exercise the second group-loop iteration, and added a dedicated FP32 test for the scalar fallback inside the WMMA launcher. The relevant tests pass.

File: comfy_kitchen/backends/hip/dlpack_bindings.cpp [hunk 2]
```diff
@@ -947,9 +949,9 @@ void gemv_awq_w4a16(nb::ndarray<> x, nb::ndarray<> qweight, nb::ndarray<> wscale
     check_hip_launch();
 }
 
-void svdquant_lora_down(nb::ndarray<> x, nb::ndarray<> lora_down, nb::ndarray<> lora_act, int M,
-                        int K, int R, uintptr_t stream_ptr) {
-    constexpr const char* kFn = "svdquant_lora_down";
+static void svdquant_lora_down_impl(const char* kFn, nb::ndarray<> x,
+                                    nb::ndarray<> lora_down, nb::ndarray<> lora_act, int M, int K,
+                                    int R, uintptr_t stream_ptr, bool use_wmma) {
```
[the remark this answers is not in the document]
Comment (tangzzycc) [reply comfy_kitchen/backends/hip/dlpack_bindings.cpp]: Thanks. The public API already makes these tensors contiguous and moves them to the correct device. I’d prefer to keep this PR focused and handle additional native binding validation separately.

[the remark this answers is not in the document]
Comment (tangzzycc) [reply comfy_kitchen/backends/hip/dlpack_bindings.cpp]: No follow-up issue is needed for now. Thanks.

**Comment by 0xDELUXA** [comment]
Ran this on RDNA4 (gfx1200). Run-verification below, plus one measurement that changes the LoRA-down half of the PR.

PR head `4ff75eb`, built with `setup.py build_ext --inplace --no-cuda --hip --hip-archs=gfx1200`:

```
pytest tests/test_svdquant_w4a4.py tests/test_hip_wmma.py -q
394 passed, 14 skipped
```

So the whole thing is run-verified on RDNA4 as well now.

## The quantizer is finished

Achieved bandwidth of `svdquant_quantize`, against a `torch` device-to-device copy on the same device as the roof (296 GB/s measured):

| dtype | M x K | time | GB/s | % of copy roof |
|---|---:|---:|---:|---:|
| fp16 | 4096 x 4096 | 0.155 ms | 274 | 92% |
| fp16 | 4096 x 14336 | 0.506 ms | 294 | 99% |
| bf16 | 4096 x 4096 | 0.155 ms | 274 | 93% |
| bf16 | 4096 x 14336 | 0.507 ms | 293 | 99% |
| fp32 | 4096 x 4096 | 0.269 ms | 284 | 96% |
| fp32 | 4096 x 14336 | 0.908 ms | 295 | 100% |

There is nothing left in that kernel. This half is a clear yes from me.

## LoRA-down is 2-4x behind `torch.mm`

The CUDA backend has no LoRA-down kernel. It computes the same product with cuBLAS, at `comfy_kitchen/backends/cuda/__init__.py:2808`:

```python
torch.mm(lora_src, lora_down, out=lora_act_rows)
```

On HIP that lands in hipBLASLt. Same tensors, same stream, 20 warmup and 100 timed iterations, min of 5 runs, bf16, with the cast included so the fp32 `lora_act` this path allocates today is preserved:

| M x K x R | `svdquant_lora_down_wmma` | `torch.mm` + cast | ratio |
|---|---:|---:|---:|
| 4096 x 4096 x 32 | 0.352 ms | 0.116 ms | 3.04x |
| 4096 x 14336 x 32 | 1.988 ms | 0.552 ms | 3.60x |
| 8192 x 14336 x 64 | 3.598 ms | 0.820 ms | 4.39x |
| 4096 x 14336 x 256 | 2.496 ms | 1.105 ms | 2.26x |
| 512 x 4096 x 32 | 0.112 ms | 0.050 ms | 2.22x |

fp16, and the variant without the cast, land in the same band: 2.0x to 10x, worst at high rank.

The 70x in the description is real, but it is 70x over a scalar kernel that was never a sensible baseline, and what replaces it is still well behind the library call the reference backend already makes for this exact product.

Where the gap comes from, reading the launch rather than a profile:

- The grid is `ceil(R / (TN * 16))` by `ceil(M / 16)` single-wave workgroups. `TN` grows with `R`, so `grid.x` stays 1 up to rank 64 and the whole GPU gets `M / 16` waves, 256 at M=4096. That number does not grow with `K`. That is why going from 4096 x 4096 x 32 to 4096 x 14336 x 32, 3.5x the K, costs 5.6x the time in my table above and 10x in the description's.
- Each lane gathers its B fragment one element at a time out of `lora_down[(kbase + i) * R + col]`, so `kFragElems` scalar loads per fragment per K step. That is 8 on gfx12 and 16 on gfx11, where both half-waves hold the same K step and therefore issue the same 16 addresses twice.

Closing both means split-K plus an LDS stage for B, which is most of a real GEMM.

So my suggestion is to drop the LoRA-down half rather than tune it: delete `lora_down_kernel` and `lora_down_wmma_kernel`, both bindings and the `has_wmma()` branch, and replace the `_C.svdquant_lora_down` call in `quantize_svdquant_w4a4` with the `torch.mm` line CUDA uses. That is faster than this PR on every shape above, removes code instead of adding it, and puts the two backends on one implementation for a computation neither of them needs a bespoke kernel for.

That leaves the quantizer, which I would take as is.

**Comment by tangzzycc** [comment]
@0xDELUXA Thanks for the detailed benchmark. Updated in 71fd9ed as suggested: I removed the custom LoRA-down kernels and bindings, and switched the HIP path to the same torch.mm implementation used by the CUDA backend. The PR now focuses on the INT4 activation quantizer optimization.

I rebuilt it on gfx1151 and reran the relevant tests: 370 passed, 14 skipped.

File: comfy_kitchen/backends/hip/__init__.py [hunk 3]
```diff
@@ -1266,9 +1266,12 @@ def quantize_svdquant_w4a4(
...
     )
-    _C.svdquant_lora_down(
-        _dl(lora_src), _dl(lora_down), _dl(lora_act[:m]), m, k, r, _stream(x)
-    )
+    if m > 0:
+        lora_act_rows = lora_act[:m]
+        if lora_act_rows.dtype == lora_src.dtype and lora_act_rows.is_contiguous():
+            torch.mm(lora_src, lora_down, out=lora_act_rows)
+        else:
+            lora_act_rows.copy_(lora_src @ lora_down, non_blocking=True)
```
[the remark this answers is not in the document]
Comment (tangzzycc) [reply comfy_kitchen/backends/hip/__init__.py:1274]: Thanks for pointing this out. This block mirrors the existing CUDA implementation, and SVDQuant is used for inference with matching operand dtypes. I’d prefer to keep the HIP and CUDA paths aligned in this PR.

**Comment by 0xDELUXA** [comment]
Could you update the title and description as well?

**Comment by tangzzycc** [comment]
@0xDELUXA  Updated the title and description to reflect the current scope. Thanks!

**Comment by tangzzycc** [comment]
@comfyanonymous ready for review when you have a moment.
